src/prefect/deployments.py

```diff
@@ -172,3 +240,195 @@ def cast_storage_to_block_type(cls, value):
+    async def upload_to_storage(self, storage_block: Block = None) -> Optional[int]:
+        """
+        Uploads the workflow this deployment represents using a provided storage block;
+        if no block is provided, defaults to configuring self for local storage.
+        """
+        deployment_path = None
+        file_count = None
+        if storage_block:
+            template = await Block.load(storage_block)
+            self.storage = template.copy(
+                exclude={"_block_document_id", "_block_document_name", "_is_anonymous"}
+            )
+
+            # upload current directory to storage location
+            file_count = await self.storage.put_directory(ignore_file=".prefectignore")
+        elif not self.storage:
+            # default storage, no need to move anything around
+            self.storage = None
+            deployment_path = str(Path(".").absolute())
+        else:
+            file_count = await self.storage.put_directory(ignore_file=".prefectignore")
+
+        # persists storage now in case it contains secret values
+        if self.storage and not self.storage._block_document_id:
+            await self.storage._save(is_anonymous=True)
+
+        self.path = deployment_path
+        return file_count
+
+    @sync_compatible
+    async def to_yaml(self, output: str):
+        """
+        Compiles the current deployment into a YAML file at the location specified by `output`.
+        """
+        with open(output, "w") as f:
+            f.write(self.header)
+            yaml.dump(self._editable_fields_dict(), f, sort_keys=False)
+            do_not_edit_msg = " DO NOT EDIT BELOW THIS LINE "
+            msg_length = len(do_not_edit_msg)
+            f.write(
+                f"###{' ' * msg_length}###\n###{do_not_edit_msg}###\n###{' ' * msg_length}###\n"
+            )
+            yaml.dump(self._immutable_fields_dict(), f, sort_keys=False)
+
+    @sync_compatible
+    async def apply(self) -> UUID:
+        """
+        Registers this deployment with the API and returns the deployment's ID.
+        """
+        if not self.name or not self.flow_name:
+            raise ValueError("Both a deployment name and flow name must be set.")
+        async with get_client() as client:
+            # prep IDs
+            flow_id = await client.create_flow_from_name(self.flow_name)
+
+            infrastructure_document_id = self.infrastructure._block_document_id
+            if not infrastructure_document_id:
+                # if not building off a block, will create an anonymous block
+                self.infrastructure = self.infrastructure.copy()
+                infrastructure_document_id = await self.infrastructure._save(
+                    is_anonymous=True,
+                )
+
+            # we assume storage was already saved
+            storage_document_id = getattr(self.storage, "_block_document_id", None)
+
+            deployment_id = await client.create_deployment(
```

Comment:
If I create a Deployment and then apply it in a script, what will happen when I run the script more than once?

Reply:
Two things:
- the project directory structure will be re-uploaded to remote storage (if remote storage is being used)
- the `create_deployment` REST route performs an upsert, so IDs, etc. remain stable

Reply:
Awesome, that's what I was hoping :)